core/src/Kokkos_SYCL.hpp

```diff
@@ -159,6 +164,7 @@ class SYCL {
 
  private:
   Impl::SYCLInternal* m_space_instance;
+  int* m_counter;
```

Comment:
`m_counter` should be a `unique_ptr<int>`.  (`m_space_instance` can't easily be one because it is sometimes but not always owned.)

Reply:
Hmm... I don't quite see how. We are trying to emulate a (poor man's version) of `std::shared_ptr` using `m_space_instance` and `m_counter` with some modifications preventing us from considering copies to the device (and thus calling finalize on the device). We use the same concept in the `Cuda` and the `HIP` backend. It might be the case that we can't copy these instances to the device in the first place (because they are not trivially copyable) but I rather err on the safe side and mirror the implementation for the other backends. Otherwise, we could, of course, just use `std::shared_ptr` (possibly with a custom deleted) and be done.

Coming back to the original issue. `m_counter` needs to be a shared state between instances and we need to know on the destructor if we are the last one. Thus, I don't believe it can be of type `std::unique_ptr`.